File: python3/unique_paths_II.py
"""
Space : O(1) 
Time  : O(n*m)

Method:
Dynamic Programming (DP)
"""


class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        m, n = len(obstacleGrid), len(obstacleGrid[0])

        for i in range(m):
            for j in range(n):
                if obstacleGrid[i][j] != 1:
                    continue

                if i == 0 and j == 0:
                    obstacleGrid[i][j] = -1
                elif i == 0:
                    obstacleGrid[i][j] = obstacleGrid[i][j-1]
                elif j == 0:
                    obstacleGrid[i][j] = obstacleGrid[i-1][j]
                else:
                    if obstacleGrid[i][j-1] != 1:
                        obstacleGrid[i][j] += obstacleGrid[i][j-1] 
                    if obstacleGrid[i-1][j] != 1:
                        obstacleGrid[i][j] += obstacleGrid[i-1][j] 

        return -obstacleGrid[-1][-1] if obstacleGrid[-1][-1] != 1 else 0

File: golang/unique_paths_II.go
package golang

// Space : O(1)
// Time  : O(n*m)

func uniquePathsWithObstacles(obstacleGrid [][]int) int {
	row := len(obstacleGrid)
	col := len(obstacleGrid[0])

	if obstacleGrid[0][0] == 1 || obstacleGrid[row-1][col-1] == 1 {
		return 0
	}

	obstacleGrid[0][0] = -1

	for y := 0; y < row; y++ {
		for x := 0; x < col; x++ {
			if obstacleGrid[y][x] != 1 {
				if 0 < y && y < row && obstacleGrid[y-1][x] < 0 {
					obstacleGrid[y][x] += obstacleGrid[y-1][x]
				}
				if 0 < x && x < col && obstacleGrid[y][x-1] < 0 {
					obstacleGrid[y][x] += obstacleGrid[y][x-1]
				}
			}
		}
	}

	return -obstacleGrid[row-1][col-1]
}
